wallet-new/src/Cardano/Wallet/Kernel/DB/TxMeta/Types.hs

```diff
@@ -0,0 +1,173 @@
+-- | Transaction metadata conform the wallet specification
+module Cardano.Wallet.Kernel.DB.TxMeta.Types (
+    -- * Transaction metadata
+    TxMeta(..)
+    -- ** Lenses
+  , txMetaId
+  , txMetaAmount
+  , txMetaInputs
+  , txMetaOutputs
+  , txMetaCreationAt
+  , txMetaIsLocal
+  , txMetaIsOutgoing
+
+  -- * Domain-specific errors
+  , TxMetaStorageError (..)
+  , InvariantViolation (..)
+
+  -- * Deep & shallow equality
+  , deepEq
+  , shallowEq
+  ) where
+
+import           Universum
+
+import           Control.Lens.TH (makeLenses)
+import qualified Data.List as List
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Text.Buildable (build)
+import           Formatting (bprint, shown, (%))
+import qualified Formatting as F
+import           Pos.Crypto (shortHashF)
+import           Serokell.Util.Text (listJsonIndent, mapBuilder)
+import           Test.QuickCheck (Arbitrary (..), Gen)
+
+import           Pos.Arbitrary.Core ()
+import qualified Pos.Core as Core
+
+{-------------------------------------------------------------------------------
+  Transaction metadata
+-------------------------------------------------------------------------------}
+
+-- | Transaction metadata
+
+--
+-- NOTE: This does /not/ live in the acid-state database (and consequently
+-- does not need a 'SafeCopy' instance), because this will grow without bound.
+data TxMeta = TxMeta {
+      -- | Transaction ID
+      _txMetaId         :: Core.TxId
+
+      -- | Total amount
+      --
+      -- TODO: What does this mean?
+    , _txMetaAmount     :: Core.Coin
+
+      -- | Transaction inputs
+    , _txMetaInputs     :: NonEmpty (Core.Address, Core.Coin)
+
+      -- | Transaction outputs
+    , _txMetaOutputs    :: NonEmpty (Core.Address, Core.Coin)
+
+      -- | Transaction creation time
+    , _txMetaCreationAt :: Core.Timestamp
+
+      -- | Is this a local transaction?
+      --
+      -- A transaction is local when /all/ of its inputs and outputs are
+      -- to and from addresses owned by this wallet.
+    , _txMetaIsLocal    :: Bool
+
+      -- | Is this an outgoing transaction?
+      --
+      -- A transaction is outgoing when it decreases the wallet's balance.
+    , _txMetaIsOutgoing :: Bool
+    }
+
+makeLenses ''TxMeta
+
+-- | Deep equality for two 'TxMeta': two 'TxMeta' are equal if they have
+-- exactly the same data, and inputs & outputs needs to appear in exactly
+-- the same order.
+deepEq :: TxMeta -> TxMeta -> Bool
```

Comment:
This naming is confusing. Usually deep versus shallow refers to how much of the substructures we compare. Here we are always comparing everything, we're just a bit more permissive. Maybe change to something like `exactlyEqualTo` and `isomorphicTo` or something like that?

Reply:
Yea, naming is hard, and I appreciate this nomenclature is a bit confusing -- your choice of names sounds good to me (bikeshed alert: I like `strictlyEqualTo` a bit better but I don't want the reader to get confused with lazy vs strict , so I won't pull the trigger on that :trollface: ).

Like `isomorphicTo` a lot -- the latin part of me approves 😀